src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -108,6 +129,122 @@ contract TokenStakingNode is
+    ) external onlyTokenStakingNodesWithdrawer returns (bytes32[] memory _fullWithdrawalRoots) {
+
+        IStrategy[] memory _strategiesArray = new IStrategy[](1);
+        _strategiesArray[0] = _strategy;
+        uint256[] memory _sharesArray = new uint256[](1);
+        _sharesArray[0] = _shares;
+        IDelegationManager.QueuedWithdrawalParams[] memory _params = new IDelegationManager.QueuedWithdrawalParams[](1);
+        _params[0] = IDelegationManager.QueuedWithdrawalParams({
+            strategies: _strategiesArray,
+            shares: _sharesArray,
+            withdrawer: address(this)
+        });
+
+        queuedShares[_strategy] += _shares;
+
+        _fullWithdrawalRoots = tokenStakingNodesManager.delegationManager().queueWithdrawals(_params);
+
+        emit QueuedWithdrawals(_strategy, _shares, _fullWithdrawalRoots);
+    }
+
+    function completeQueuedWithdrawals(
+        uint256 _nonce,
+        uint32 _startBlock,
+        uint256 _shares,
+        IStrategy _strategy,
+        uint256[] memory _middlewareTimesIndexes
+    ) public onlyTokenStakingNodesWithdrawer {
+
+        IDelegationManager _delegationManager = tokenStakingNodesManager.delegationManager();
+
+        IDelegationManager.Withdrawal[] memory _withdrawals = new IDelegationManager.Withdrawal[](1);
+        {
+            IStrategy[] memory _strategiesArray = new IStrategy[](1);
+            _strategiesArray[0] = _strategy;
+            uint256[] memory _sharesArray = new uint256[](1);
+            _sharesArray[0] = _shares;
+            _withdrawals[0] = IDelegationManager.Withdrawal({
+                staker: address(this),
+                delegatedTo: _delegationManager.delegatedTo(address(this)),
+                withdrawer: address(this),
+                nonce: _nonce,
+                startBlock: _startBlock,
+                strategies: _strategiesArray,
+                shares: _sharesArray
+            });
+        }
+
+        uint256 _expectedAmountOut = _strategy.sharesToUnderlyingView(_shares);
+
+        IERC20 _token = _strategy.underlyingToken();
+        uint256 _balanceBefore = _token.balanceOf(address(this));
+
+        {
+            bool[] memory _receiveAsTokens = new bool[](1);
+            _receiveAsTokens[0] = true;
+            IERC20[][] memory _tokens = new IERC20[][](1);
+            _tokens[0] = new IERC20[](1);
+            _tokens[0][0] = _token;
+
+            _delegationManager.completeQueuedWithdrawals(
+                _withdrawals,
+                _tokens,
+                _middlewareTimesIndexes,
+                _receiveAsTokens
+            );
+        }
+
+        uint256 _actualAmountOut = _token.balanceOf(address(this)) - _balanceBefore;
+        uint256 _delta = _actualAmountOut > _expectedAmountOut ?
+            _actualAmountOut - _expectedAmountOut :
+            _expectedAmountOut - _actualAmountOut;
+        if (_delta > 2) revert WithdrawalAmountMismatch(); // @todo - might be a footgun
```

Comment:
why not exactly equal?

is it because of stETH and the 2 wei error problem?

https://github.com/lidofinance/lido-dao/issues/442

Reply:
exactly